File: editor/en/[85]Maximal Rectangle.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        rows = len(matrix)
        cols = len(matrix[0])
        max_area = 0
        height = [0] * cols

        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == '0':
                    height[j] = 0
                else:
                    height[j] += 1
            max_area = max(max_area, self.largest_rectangle_area(height))

        return max_area

    def largest_rectangle_area(self, heights: List[int]) -> int:
        stack = [-1]
        max_area = 0
        heights.append(0)

        for i in range(len(heights)):
            while stack[-1] != -1 and heights[stack[-1]] > heights[i]:
                h = heights[stack.pop()]
                w = i - stack[-1] - 1
                max_area = max(max_area, h * w)
            stack.append(i)

        heights.pop()
        return max_area

# leetcode submit region end(Prohibit modification and deletion)

File: editor/en/leetcode/editor/en/[85]Maximal Rectangle.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int maximalRectangle(char[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int maxArea = 0;
        int[] height = new int[cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == '0') {
                    height[j] = 0;
                } else {
                    height[j]++;
                }
            }
            maxArea = Math.max(maxArea, largestRectangleArea(height));
        }

        return maxArea;
    }

    private int largestRectangleArea(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(-1);
        int maxArea = 0;

        for (int index = 0; index < heights.length; index++) {
            while (stack.peek() != -1 && heights[stack.peek()] > heights[index]) {
                int h = heights[stack.pop()];
                int w = index - stack.peek() - 1;
                maxArea = Math.max(maxArea, h * w);
            }

            stack.push(index);
        }

        while (stack.peek() != -1) {
            int h = heights[stack.pop()];
            int w = heights.length - stack.peek() - 1;
            maxArea = Math.max(maxArea, h * w);
        }

        return maxArea;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
